Java Lambda tracing installation steps (#9902)
DataDog/documentation: content/en/serverless/installation/java.md

  text: 'Submitting Custom Metrics from Serverless Applications'
---

{{< img src="serverless/java-lambda-tracing.png" alt="Monitor Java Lambda Functions with Datadog"  style="width:100%;">}}

## Required setup

If not already configured:

- Install the [AWS integration][1]. This allows Datadog to ingest Lambda metrics from AWS. 
- Install the [Datadog Forwarder Lambda function][2], which is required to ingest AWS Lambda traces, enhanced metrics, custom metrics, and logs. 

After you have installed the [AWS integration][1] and the [Datadog Forwarder][2], follow these steps to instrument your application to send [enhanced Lambda metrics][3], logs, and traces to Datadog. 

> To fully instrument your serverless application with distributed tracing, your Java Lambda functions must be using the Java 8 Corretto (`Java8.al2`) or Java 11 (`Java11.al2`) runtimes.

Reviewer: ```suggestion
To fully instrument your serverless application with distributed tracing, your Java Lambda functions must be using the Java 8 Corretto (`java8.al2`) or Java 11 (`java11`) runtimes.
```

You need to subscribe the Datadog Forwarder Lambda function to each of your function’s log groups, in order to send metrics, traces and logs to Datadog.

1. [Install the Datadog Forwarder if you haven't][2].
2. [Subscribe the Datadog Forwarder to your function's log groups][6].

### Monitor Java Lambda function cold starts

> Cold starts occur when your serverless applications receive sudden increases in traffic, and can occur when the function was previously inactive or when it was receiving a relatively constant number of requests. Users may perceive cold starts as slow response times or lag. Datadog highly recommends you configure a monitor on Java Lambda function cold starts, and use Datadog Serverless Insights to [keep cold starts to a minimum][7].

Reviewer 2: ```suggestion
Cold starts occur when your serverless applications receive sudden increases in traffic, including when the function was previously inactive or when it was receiving a relatively constant number of requests. Users may perceive cold starts as slow response times or lag. Datadog highly recommends you configure a monitor on Java Lambda function cold starts, and use Datadog Serverless Insights to [keep cold starts to a minimum][7].
```

{{< img src="serverless/java-monitor-cold-starts.png" alt="Monitor Java Lambda Function Cold Starts"  style="width:100%;">}}

> To create a Datadog monitor on Java Lambda function cold starts, follow the monitor creation steps [here][8] with the following criteria:

Reviewer 2: ```suggestion
To create a Datadog monitor on Java Lambda function cold starts, follow the [monitor creation steps][8] with the following criteria:
```

- Metric Name: `aws.lambda.enhanced.invocations`
- From: `runtime:java*` and `cold_start:true`
- Alert Grouping: Multi Alert, trigger a separate alert for each `function_arn`

### Unified service tagging

Although it's optional, Datadog highly recommends tagging you serverless applications with the `env`, `service`, and `version` tags following the [unified service tagging documentation][9].

## Explore Datadog serverless monitoring

After you have configured your function following the steps above, you should be able to view metrics, logs and traces on the [Serverless Homepage][10].

> ### Monitor custom business logic

Reviewer: Should this be a sub-heading under "Explore Datadog Serverless Monitoring", or at an even level?

Author: It's intentional, I'm actually surprised it wasn't already a sub-heading :)

        URL url = new URL("https://example.com");
        HttpURLConnection hc = (HttpURLConnection)url.openConnection();
        hc.connect();

        ddl.finish();
    }
}
```

> For more information on custom metric submission, see [here][11].

Reviewer 2: ```suggestion
See the [custom metrics documentation][11] for more information on custom metric submission.
```

### Connect logs and traces

> Follow the instructions [here][12] to automatically connect Java Lambda function logs and traces.

Reviewer 2: ```suggestion
To automatically connect Java Lambda function logs and traces, see the [Connecting Java Logs and Traces][12] for instructions.
```

> To automatically connect Java Lambda function logs and traces, see the [Connecting Java Logs and Traces][12] for instructions.

Author: ```suggestion
To automatically connect Java Lambda function logs and traces, see [Connecting Java Logs and Traces][12] for instructions.
```
